src/MOI_wrapper.jl

```diff
@@ -0,0 +1,286 @@
+            error("No starting value specified for `$i`th variable.")
+        else
+            model.variable_primal_start[i]
+        end for i in eachindex(model.variable_primal_start)
+    ]
+    function eval_f_cb(M, x)
+        obj = MOI.eval_objective(model, x)
+        if model.sense == MOI.MAX_SENSE
+            obj = -obj
+        end
+        return obj
+    end
+    function eval_grad_f_cb(M, x)
+        grad_f = zeros(length(x))
+        MOI.eval_objective_gradient(model, grad_f, x)
+        if model.sense == MOI.MAX_SENSE
+            LinearAlgebra.rmul!(grad_f, -1)
+        end
+        reshaped_grad_f = JuMP.reshape_vector(grad_f, _shape(model.manifold))
+        return ManifoldDiff.riemannian_gradient(model.manifold, x, reshaped_grad_f)
+    end
+    MOI.initialize(model.nlp_data.evaluator, [:Grad])
+    mgo = Manopt.ManifoldGradientObjective(eval_f_cb, eval_grad_f_cb)
+    dmgo = decorate_objective!(model.manifold, mgo)
+    model.problem = DefaultManoptProblem(model.manifold, dmgo)
+    reshaped_start = JuMP.reshape_vector(start, _shape(model.manifold))
+    s = GradientDescentState(model.manifold, reshaped_start)
+    model.state = decorate_state!(s)
+    solve!(model.problem, model.state)
+    return nothing
+end
+
+using JuMP: JuMP
+
+struct ArrayShape{N} <: JuMP.AbstractShape
+    size::NTuple{N,Int}
+end
+
+function JuMP.vectorize(array::Array{T,N}, ::ArrayShape{M}) where {T,N,M}
+    return vec(array)
+end
+
+function JuMP.reshape_vector(vector::Vector, shape::ArrayShape)
+    return reshape(vector, shape.size)
+end
+
+function _shape(m::ManifoldsBase.AbstractManifold)
+    return ArrayShape(ManifoldsBase.representation_size(m))
+end
+
+function JuMP.build_variable(::Function, func, m::ManifoldsBase.AbstractManifold)
+    shape = _shape(m)
+    return JuMP.VariablesConstrainedOnCreation(
+        JuMP.vectorize(func, shape), VectorizedManifold(m), shape
+    )
+end
+
+function MOI.get(model::Optimizer, ::MOI.TerminationStatus)
+    if isnothing(model.state)
+        return MOI.OPTIMIZE_NOT_CALLED
+    else
+        return MOI.LOCALLY_SOLVED
+    end
+end
+
+function MOI.get(model::Optimizer, ::MOI.ResultCount)
+    if isnothing(model.state)
+        return 0
+    else
+        return 1
+    end
+end
+
+function MOI.get(model::Optimizer, ::MOI.PrimalStatus)
+    if isnothing(model.state)
+        return MOI.NO_SOLUTION
+    else
+        return MOI.FEASIBLE_POINT
+    end
+end
+
+MOI.get(::Optimizer, ::MOI.DualStatus) = MOI.NO_SOLUTION
+
+function MOI.get(model::Optimizer, ::MOI.RawStatusString)
+    # `strip` removes the `\n` at the end and returns an `AbstractString`
+    # Since MOI wants a `String`, we pass it through `string`
+    return string(strip(get_reason(model.state)))
+end
+
+function MOI.get(model::Optimizer, attr::MOI.ObjectiveValue)
+    MOI.check_result_index_bounds(model, attr)
+    solution = get_solver_return(get_objective(model.problem), model.state)
```

Comment:
```suggestion
    solution = get_solver_return(model.state)
```

The 2-element solver return function (and the one with a tuple) are just for ease of use. To be precise: A solver can return just `x`, just the state, the objective and `x` or the objective a the state.

They all yield different displays of the results, for example to see the Cache usage, the objective has to be present.

But to access the solvers result, you always just need the state :)

Reply:
I see, thanks :)